{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Configure the model architecture and training parameters\n",
    "\n",
    "This tutorial provides a step-by-step guide on configuring the model architectures, training hyperparameters, and analysis of time-series single dataset using UNAGI. We demonstrate the capabilities of UNAGI by applying it to scRNA-seq data sampled from a single-nuclei RNA sequencing data.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from UNAGI import UNAGI\n",
    "unagi = UNAGI()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: Setup and load the datasets\n",
    "\n",
    "After loading UNAGI package, we need to setup the data for UNAGI training.\n",
    "\n",
    "-   We need to specify the data path of your h5ad files after stage segmentation. e.g. '../data/small/0.h5ad'. Then UNAGI will load all h5ad files in the target directory. \n",
    "\n",
    "-   UNAGI requires the total number of time-points the dataset has as the input. e.g. total_stage=4\n",
    "\n",
    "-   UNAGI requires the key of time-points attribute in the annData.obs table.\n",
    "\n",
    "-   If the dataset is not splited into individual stages, you can specify the splited_dataset as False to segment the dataset.\n",
    "\n",
    "-   To build the K-Nearest Neighbors (KNN) connectivity matrix in Graph convolution training, the neighbors number of KNN should be defined. The default value is 25. \n",
    "\n",
    "-   You can also specify how many threads you want to use when using UNAGI. The default number of threads is 20. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘../UNAGI/data/example/0’: File exists\n"
     ]
    }
   ],
   "source": [
    "unagi.setup_data('../UNAGI/data/example',total_stage=4,stage_key='stage')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Configure the model architecture of UNAGI and training hyper-parameters\n",
    "\n",
    "First, it's mandatory to specify the **task** your are executing. (e.g. we call the example dataset as task='small_sample') The **task** is the identifier of your experiments and you can reterive the trained model and the results of each iteration at '../data/**task**/' directory. \n",
    "\n",
    "Next, you will need to specify the distribution of you single cell data. UNAGI provides negative binomial (NB), zero-inflated negative binomial (ZINB), zero-inflated log normal, and normal distribution to model your single cell data.\n",
    "\n",
    "You can use the *device* keyword to specify the device you want to use for training.\n",
    "\n",
    "'epoch_initial': the number of training epochs for the first iteration.\n",
    "\n",
    "'epoch_iter': the number of training epochs for the iterative training.\n",
    "\n",
    "'max_iter': the total number of iterations UNAGI will run\n",
    "\n",
    "'BATCHSIZE': the batch size of a mini-batch\n",
    "\n",
    "'lr': the learning rate of Graph VAE\n",
    "\n",
    "'lr_dis': the learning rate of the adversarial discriminator\n",
    "\n",
    "'latent_dim': the dimension of Z space\n",
    "\n",
    "'hiddem_dim': the neuron size of each fully connected layers\n",
    "\n",
    "'graph_dim': the dimension of graph representation\n",
    "\n",
    "After settingt the training hyper parameters and model architectures, you can use `unagi.run_UNAGI()` to start training. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unagi.setup_training(task='example',dist='ziln',device='cuda:0',GPU=True,epoch_iter=5,epoch_initial=2,max_iter=3,BATCHSIZE=560)\n",
    "unagi.run_UNAGI(idrem_dir = '../../idrem',CPO=True) #specify the path to the idrem (installed separately) directory, set CPO to True if you want to use CPO for the training (extra time needed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 3: Perform in-silico perturbations and downstream analysis\n",
    "\n",
    "After training the UNAGI model, you can perfrom downstream tasks including hierarchical static marker discovries \n",
    "parameters: \n",
    "data_path: the directory of the dataset generated by UNAGI\n",
    "iteration: the iteration of the dataset belongs to\n",
    "progressionmarker_background_sampling_times: the number of sampling times to generate the dynamic marker backgrounds\n",
    "target_dir: the directory to store the downstream analysis results and h5ad files\n",
    "customized_drug: the directory to customized drug profile\n",
    "cmap_dir: the directory to the precomputed CMAP database which contains the drug/compounds and their regualted genes and regualated directions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from UNAGI import UNAGI\n",
    "unagi = UNAGI()\n",
    "unagi.analyse_UNAGI('../UNAGI/data/example/2/stagedata/org_dataset.h5ad',2,10,target_dir=None,customized_drug='../UNAGI/data/jasper_target_pair.npy',cmap_dir='../../CMAPDirectionDf.npy')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wks",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
