{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test\n",
    "import os\n",
    "idrem_dir = '/mnt/md0/yumin/to_upload/idrem'\n",
    "trained = False\n",
    "species= 'Human'\n",
    "examplefile_path = os.path.join(idrem_dir, 'example_settings.txt')#open(os.path.join(idrem_dir, 'example_settings.txt'), 'r')\n",
    "settings_file_path = 'test.txt'#os.path.join(midpath, str(iteration), 'idremsetting', f'{file_name}.txt')\n",
    "with open(examplefile_path, 'r') as examplefile:\n",
    "# Open settings_file_path for writing\n",
    "    with open(settings_file_path, 'w') as f:\n",
    "        for k, line in enumerate(examplefile.readlines()):\n",
    "\n",
    "            if trained and k == 4:\n",
    "                print(midpath) \n",
    "                f.write('%s\\t%s\\n' % ('Saved_Model_File', os.path.join(os.path.abspath(os.path.join(midpath, str(iteration), 'idremInput')), f'{file_name}.txt')))\n",
    "            elif k == 1:\n",
    "                if species == 'Human':\n",
    "                    f.write('%s\\t%s\\n' % ('TF-gene_Interaction_Source', 'human_encode.txt.gz'))\n",
    "                    continue\n",
    "            elif k == 2:\n",
    "                if species == 'Human':\n",
    "                    f.write('%s\\t%s\\n' % ('TF-gene_Interactions_File', 'TFInput/human_encode.txt.gz'))\n",
    "                    continue\n",
    "            elif k == 5:\n",
    "                if species == 'Human':\n",
    "                    f.write('%s\\t%s\\n' % ('Gene_Annotation_Source', 'Human (EBI)'))\n",
    "                    continue\n",
    "            elif k == 6:\n",
    "                if species == 'Human':\n",
    "                    f.write('%s\\t%s\\n' % ('Gene_Annotation_File', 'goa_human.gaf.gz'))\n",
    "                    continue \n",
    "            elif k== 17:\n",
    "                f.write('%s\\n' % ('miRNA-gene_Interaction_Source'))\n",
    "                continue\n",
    "            elif k== 18:\n",
    "                f.write('%s\\n' % ('miRNA_Expression_Data_File'))\n",
    "                continue\n",
    "            elif k== 26:\n",
    "                f.write('%s\\n' % ('Proteomics_File'))\n",
    "                continue\n",
    "            elif k == 34:\n",
    "                f.write('%s\\n' % ('Epigenomic_File'))\n",
    "                continue\n",
    "            elif k == 35:\n",
    "                f.write('%s\\n' % ('GTF File'))\n",
    "                continue\n",
    "            elif k == 42:\n",
    "                f.write('%s\\t%s\\n' % ('Minimum_Absolute_Log_Ratio_Expression', str(Minimum_Absolute_Log_Ratio_Expression)))\n",
    "                continue\n",
    "            elif k ==51 :\n",
    "                f.write('%s\\t%s\\n' % ('Convergence_Likelihood_%', str(Convergence_Likelihood)))\n",
    "                continue\n",
    "            elif k == 52:\n",
    "                f.write('%s\\t%s\\n' % ('Minimum_Standard_Deviation', str(Minimum_Standard_Deviation)))\n",
    "                continue\n",
    "            elif k != 3:\n",
    "                f.write(line)\n",
    "            else:\n",
    "                f.write('%s\\t%s\\n' % ('Expression_Data_File', os.path.join(os.path.abspath(os.path.join(midpath, str(iteration), 'idremInput')), f'{file_name}.txt')))\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
